packages/tools/client-debugger/client-debugger-view/src/Debugger.tsx

```diff
@@ -40,82 +54,98 @@ export interface FluidClientDebuggersProps {
+		 */
+		const inboundMessageHandlers: InboundHandlers = {
+			["REGISTRY_CHANGE"]: (untypedMessage) => {
+				const message = untypedMessage as RegistryChangeMessage;
+				setContainers(message.data.containers);
+				return true;
+			},
+		};
 
-	function getDebuggerFromContainerId(containerId: string): IFluidClientDebugger {
-		const match = clientDebuggers.find(
-			(clientDebugger) => clientDebugger.containerId === containerId,
-		);
-		if (match === undefined) {
-			throw new Error(`No debugger found associated with Container ID "${containerId}".`);
+		/**
+		 * Event handler for messages coming from the Message Relay
+		 */
+		function messageHandler(message: Partial<IDebuggerMessage>): void {
+			handleIncomingMessage(message, inboundMessageHandlers, {
+				context: loggingContext,
+			});
 		}
-		return match;
-	}
 
-	React.useEffect(() => {
-		function onDebuggerChanged(): void {
-			const newDebuggerList = getFluidClientDebuggers();
-			setClientDebuggers(newDebuggerList);
-			if (selectedContainerId === undefined) {
-				const newSelection = getDefaultDebuggerSelectionId(newDebuggerList);
-				console.log(`Updating selection to container ID "${newSelection}".`);
-				setSelectedContainerId(newSelection);
-			}
-		}
+		messageRelay.on("message", messageHandler);
 
-		debuggerRegistry.on("debuggerRegistered", onDebuggerChanged);
-		debuggerRegistry.on("debuggerClosed", onDebuggerChanged);
+		messageRelay.postMessage(getContainerListMessage);
 
 		return (): void => {
-			debuggerRegistry.off("debuggerRegistered", onDebuggerChanged);
-			debuggerRegistry.off("debuggerClosed", onDebuggerChanged);
+			messageRelay.off("message", messageHandler);
 		};
-	}, [getDefaultDebuggerSelectionId, selectedContainerId, debuggerRegistry, setClientDebuggers]);
-
-	const view =
-		selectedContainerId === undefined ? (
-			<NoDebuggerInstance
-				onRetryDebugger={(): void => {
-					const newDebuggerList = getFluidClientDebuggers();
-					setClientDebuggers(newDebuggerList);
-					const newDefaultId = getDefaultDebuggerSelectionId(newDebuggerList);
-					setSelectedContainerId(newDefaultId);
-				}}
-			/>
-		) : (
-			<ClientDebugView
-				clientDebugger={getDebuggerFromContainerId(selectedContainerId)}
-				renderOptions={props.renderOptions}
-			/>
-		);
-
-	const selectionView: React.ReactElement =
-		clientDebuggers.length > 1 ? (
-			<ContainerSelectionDropdown
-				initialSelection={selectedContainerId}
-				options={clientDebuggers.map((clientDebugger) => ({
-					id: clientDebugger.containerId,
-					nickname: clientDebugger.containerNickname,
-				}))}
-				onChangeSelection={(containerId): void => setSelectedContainerId(containerId)}
-			/>
-		) : (
-			<></>
-		);
+	}, [setContainers, messageRelay]);
+
+	let innerView: React.ReactElement;
+	switch (menuSelection) {
+		case "telemetry":
+			innerView = <TelemetryView />;
+			break;
+		case "container":
+			// eslint-disable-next-line no-case-declarations
```

Comment:
We should probably hoist the responsibility of container selection out of this component and into its own. Single-reponsibility (state-wise) components are the way to go in my opinion. If this component is responsible for managing the outer menu selection (containers vs telemetry vs etc.), then a different component should manage the `containers` view and any state it requires.